File: backend/app/schemas/pm.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class PmRegisterIn(BaseModel):
    email: str
    password: str = Field(min_length=4)
    display_name: str = ""


class PmLoginIn(BaseModel):
    email: str
    password: str


class PmWorkspaceSummary(ORMModel):
    id: UUID
    name: str
    slug: str
    role: str
    can_create_features: bool = False


class PmUserOut(ORMModel):
    id: UUID
    email: str
    display_name: str
    created_at: datetime
    workspaces: list[PmWorkspaceSummary] = Field(default_factory=list)


class PmWorkspaceIn(BaseModel):
    name: str
    slug: str = ""


class PmMemberIn(BaseModel):
    email: str
    role: str = "member"
    display_name: str = ""
    password: str = ""


class PmMemberUpdate(BaseModel):
    role: str | None = None
    display_name: str | None = None


class PmMemberOut(ORMModel):
    id: UUID
    user_id: UUID
    email: str
    display_name: str
    role: str
    online: bool = False
    last_seen_at: datetime | None = None
    created_at: datetime
    temporary_password: str | None = None


class PmWorkspaceOut(ORMModel):
    id: UUID
    name: str
    slug: str
    role: str
    can_create_features: bool = False
    created_at: datetime
    members: list[PmMemberOut] = Field(default_factory=list)


class PmProjectIn(BaseModel):
    name: str
    key: str = ""
    description: str = ""


class PmProjectUpdate(BaseModel):
    name: str | None = None
    description: str | None = None


class PmProjectOut(ORMModel):
    id: UUID
    workspace_id: UUID
    name: str
    key: str
    description: str
    sweety_brand_id: UUID | None
    created_at: datetime
    feature_count: int = 0
    issue_count: int = 0


class PmFeatureIn(BaseModel):
    title: str
    description: str = ""
    status: str = "backlog"
    stories: list[str] = Field(default_factory=list)


class PmFeatureUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: str | None = None


class PmFeatureOut(ORMModel):
    id: UUID
    project_id: UUID
    key: str
    number: int
    title: str
    description: str
    status: str
    source: str
    created_by_user_id: UUID | None
    issue_count: int = 0
    created_at: datetime
    updated_at: datetime


class PmIssueIn(BaseModel):
    title: str
    description: str = ""
    kind: str = "task"
    status: str = "backlog"
    priority: int = 2
    feature_id: UUID | None = None
    parent_id: UUID | None = None
    assignee_id: UUID | None = None
    due_date: date | None = None


class PmIssueUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    kind: str | None = None
    status: str | None = None
    priority: int | None = None
    feature_id: UUID | None = None
    parent_id: UUID | None = None
    assignee_id: UUID | None = None
    due_date: date | None = None
    sort_order: int | None = None


class PmIssueOut(ORMModel):
    id: UUID
    project_id: UUID
    feature_id: UUID | None
    parent_id: UUID | None = None
    parent_key: str = ""
    subticket_count: int = 0
    key: str
    number: int
    title: str
    description: str
    kind: str
    status: str
    priority: int
    assignee_id: UUID | None
    reporter_id: UUID | None
    assignee_name: str = ""
    reporter_name: str = ""
    feature_title: str = ""
    feature_key: str = ""
    due_date: date | None
    overdue: bool = False
    sort_order: int
    created_at: datetime
    updated_at: datetime


class PmCommentIn(BaseModel):
    body: str


class PmCommentOut(ORMModel):
    id: UUID
    issue_id: UUID
    user_id: UUID
    author_name: str = ""
    body: str
    created_at: datetime


class PmBoardOut(BaseModel):
    project: PmProjectOut
    features: list[PmFeatureOut]
    columns: dict[str, list[PmIssueOut]]
    members: list[PmMemberOut]
    role: str = "member"
    can_create_features: bool = False


class PmChatIn(BaseModel):
    content: str


class PmWorkspaceUpdate(BaseModel):
    name: str | None = None


class PmCommentUpdate(BaseModel):
    body: str


class PmReportOut(BaseModel):
    totals: dict
    by_status: dict[str, int]
    by_kind: dict[str, int]
    by_assignee: list[dict]
    by_epic: list[dict]
    overdue: list[PmIssueOut]

File: frontend/lib/pm.ts
export const PM_TOKEN_KEY = "sweety_pm_token";
export const PM_WORKSPACE_KEY = "sweety_pm_workspace";

export function getPmToken(): string | null {
  if (typeof window === "undefined") return null;
  return localStorage.getItem(PM_TOKEN_KEY);
}

export function setPmToken(token: string) {
  localStorage.setItem(PM_TOKEN_KEY, token);
}

export function clearPmToken() {
  localStorage.removeItem(PM_TOKEN_KEY);
}

export function getPmWorkspaceId(): string | null {
  if (typeof window === "undefined") return null;
  return localStorage.getItem(PM_WORKSPACE_KEY);
}

export function setPmWorkspaceId(id: string) {
  localStorage.setItem(PM_WORKSPACE_KEY, id);
}

export async function pmApi<T>(path: string, options: RequestInit = {}): Promise<T> {
  const token = getPmToken();
  const headers: Record<string, string> = {
    ...(options.body ? { "Content-Type": "application/json" } : {}),
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
    ...(options.headers as Record<string, string> | undefined),
  };
  const res = await fetch(`/api/v1/pm${path}`, {
    ...options,
    headers,
    signal: options.signal ?? (typeof AbortSignal.timeout === "function" ? AbortSignal.timeout(25_000) : undefined),
  });
  if (res.status === 401 && typeof window !== "undefined") {
    clearPmToken();
    if (!path.startsWith("/auth/")) window.location.href = "/pm/login";
  }
  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || res.statusText);
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

export type PmWorkspace = {
  id: string;
  name: string;
  slug: string;
  role: string;
  can_create_features: boolean;
  created_at: string;
  members: PmMember[];
};

export type PmMember = {
  id: string;
  user_id: string;
  email: string;
  display_name: string;
  role: string;
  created_at: string;
};

export type PmMe = {
  id: string;
  email: string;
  display_name: string;
  created_at: string;
  workspaces: {
    id: string;
    name: string;
    slug: string;
    role: string;
    can_create_features: boolean;
  }[];
};

export type PmProject = {
  id: string;
  workspace_id: string;
  name: string;
  key: string;
  description: string;
  sweety_brand_id: string | null;
  created_at: string;
  feature_count: number;
  issue_count: number;
};

export type PmFeature = {
  id: string;
  project_id: string;
  key: string;
  number: number;
  title: string;
  description: string;
  status: string;
  source: string;
  created_by_user_id: string | null;
  issue_count: number;
  created_at: string;
  updated_at: string;
};

export type PmIssue = {
  id: string;
  project_id: string;
  feature_id: string | null;
  key: string;
  number: number;
  title: string;
  description: string;
  kind: string;
  status: string;
  priority: number;
  assignee_id: string | null;
  reporter_id: string | null;
  assignee_name: string;
  reporter_name: string;
  feature_title: string;
  feature_key: string;
  due_date: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type PmComment = {
  id: string;
  issue_id: string;
  user_id: string;
  author_name: string;
  body: string;
  created_at: string;
};

export type PmBoard = {
  project: PmProject;
  features: PmFeature[];
  columns: Record<string, PmIssue[]>;
  members: PmMember[];
  role: string;
  can_create_features: boolean;
};

export type PmChatMessage = {
  id: string;
  role: string;
  content: string;
  trace?: unknown[];
  created_at: string | null;
};

export const ISSUE_COLUMNS = ["backlog", "todo", "in_progress", "review", "done"] as const;
export const ISSUE_KINDS = ["story", "task", "bug"] as const;
export const FEATURE_STATUSES = ["backlog", "planned", "in_progress", "done"] as const;
export const PRIORITY_LABELS = ["Urgent", "High", "Medium", "Low", "None"];

export function priorityLabel(n: number) {
  return PRIORITY_LABELS[n] || "Medium";
}
